File: ws-fed_cfstest.js
const express = require('express');
const { parseString } = require('xml2js');
const fs = require('fs');
const app = express();

app.use(express.urlencoded({ extended: true }));
app.use(express.static('public'));

app.set('view engine', 'ejs');

app.listen(3000, () => {
  console.log('Server is running on port 3000');
});

function readIdpMetadata() {
  const metadataXml = fs.readFileSync('ws-fed_cfstest.xml', 'utf-8');
  return metadataXml;
}

const idpMetadataXml = readIdpMetadata();

const idpMetadata = {
  //entity_id: 'https://cfs001.dudleycadet.us',
  entity_id: 'https://cfs-server',
  sso_service: {
    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST',
    //url: 'https://cfs001.dudleycadet.us/cfs/WsFed/testusers/9bc41285-8afc-40a0-ad12-22e0d0d09197'
    url: 'https://cfs-server/cfs/WsFed/krpocthree/e01d5c4b-2f12-4414-88f6-d0449323d8b1'
  }
};

app.get('/', (req, res) => {
  if (req.session.user) {
    res.send(`Hello, ${req.session.user.name}! <a href="/logout">Logout</a>`);
  } else {
    res.send('Hello, guest! <a href="/login">Login with WS-Federation</a>');
  }
});

app.get('/login', (req, res) => {
  res.render('login');
});

app.post('/login', (req, res) => {
  const email = req.body.email;
  const password = req.body.password;

  // Validate the user's credentials
  if (users[email] && users[email].password === password) {
    // Create and send WS-Federation request
    res.redirect(`/wsfed?email=${email}`);
  } else {
    res.send('Invalid credentials');
  }
});

app.get('/wsfed', (req, res) => {
  const email = req.query.email;
  const idpLoginUrl = idpMetadata.sso_service.url;

  const encodedRedirectUrl = encodeURIComponent('http://localhost:3000/wsfed-callback');

  // Construct the WS-Federation login URL
  const wsfedLoginUrl = `${idpLoginUrl}?whr=${idpMetadata.entity_id}&wreply=${encodedRedirectUrl}&username=${email}`;

  res.redirect(wsfedLoginUrl);
});

app.get('/wsfed-callback', (req, res) => {
  // Handle WS-Federation response
  // ...
});

app.get('/logout', (req, res) => {
  req.session.user = null;
  res.redirect('/');
});

const users = {
  'testuser001@krpoc.com': { name: 'TestUser001', password: 'P@ssword123' },
  'testuser001@krpoc.com': { name: 'TestUser001', password: 'P@ssword123' },
  'testuser002@krpoc.com': { name: 'TestUser002', password: 'P@ssword123' }
};

File: cgi-bin/ws-fed_cfstest.py
#!/usr/bin/python3
from flask import Flask, redirect, request, session, url_for
from saml2 import BINDING_HTTP_POST
from saml2.client import Saml2Client

app = Flask(__name__)
app.secret_key = 'your-secret-key'

# IdP configuration
idp_metadata = {
    "entity_id": "https://cfs001.dudleycadet.us",
    "sso_service": {
        "binding": BINDING_HTTP_POST,
        "url": "https://cfs001.dudleycadet.us/cfs/WsFed/testusers/9bc41285-8afc-40a0-ad12-22e0d0d09197"
    }
}

# Simulated user database
users = {
    'testuser001@dudleycadet.us': {'name': 'TestUser001', 'password': 'My2Kids000'},
    'testuser002@dudleycadet.us': {'name': 'TestUser002', 'password': 'My2Kids000'}
}

@app.route('/')
def index():
    if 'user' in session:
        return f'Hello, {session["user"]["name"]}! <a href="/logout">Logout</a>'
    return 'Hello, guest! <a href="/login">Login with WS-Federation</a>'

@app.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        email = request.form['email']
        password = request.form['password']

        # Validate the user's credentials
        if email in users and users[email]['password'] == password:
            client = Saml2Client(None)
            authn_request = client.create_authn_request(idp_metadata)
            redirect_url = authn_request
            return redirect(redirect_url)
        else:
            return 'Invalid credentials'

    return '''
    <form method="post">
        <label for="email">Email:</label>
        <input type="email" id="email" name="email" required><br><br>
        <label for="password">Password:</label>
        <input type="password" id="password" name="password" required><br><br>
        <button type="submit">Login</button>
    </form>
    '''

@app.route('/acs', methods=['POST'])
def acs():
    # Process the assertion response from IdP
    pass

@app.route('/logout')
def logout():
    session.pop('user', None)
    return redirect(url_for('index'))

if __name__ == '__main__':
    app.run()
